### ISSUE TITLE
Implementing "Smart-Casting"

### ISSUE THREAD

**Issue by Sembrani** [loopuleasa/Dota2-TheCore-Config-Engine#58]
As mentioned in #26, how would you like this accessed? Have it affected by a toggle so that you have either smart-cast or quickcast? or have it as a third toggle option to cycle between?

**Comment by loopuleasa** [comment]
probably as a toggle, but first I think as a concept you should try implementing smart-cast independent (in the new upcoming "tutorials" folder)

**Comment by Sembrani** [comment]
Sure. I have some forks for trialing things now, so it should go fairly smoothly.

**Comment by loopuleasa** [comment]
BTW, smart-casting shouldn't be a priority since without sixense_left_click on reborn, this function is exclusive only to the Legacy client, and I am getting tired of branching the development in two places.

**Comment by loopuleasa** [comment]
In reality, both should tend to be the same.

**Comment by loopuleasa** [comment]
Smart-casting discussion on dev dota http://dev.dota2.com/showthread.php?t=173103

**Comment by Sembrani** [comment]
It should still be able to be implemented in reborn. Use +alias "dota_ability_execute", then -alias "dota_ability_quickcast" ?

**Comment by loopuleasa** [comment]
interesting idea!

**Comment by loopuleasa** [comment]
I think the Quick/Normal mode along with the ALT+SPACE+Key toggles should switch between these 4 modes:

- quick primary, normal secondary
- normal primary, quick secondary
- quick primary, smart secondary
- smart primary, quick secondary

and I would love if pressing/depressing the space key would cancel smart casts (if you want your character to stop the action). Hope you understand what I mean.

**Comment by leodag** [comment]
That'd be pretty easy;  just

alias nothing ""
alias quickcast0 "dota_ability_quickcast 0"
alias +smartcast0 "dota_ability_execute 0; alias -smartcast0 quickcast0"

Then, at the end of dota2_keybinds_space_pressed.cfg put a
alias -smartcast0 nothing

Might be a bit sloppy since I don't actually use your config, but it'd be easy to adapt.

**Comment by loopuleasa** [comment]
Yep, that's how it's going to be in the end. 
Haven't had much time to work on this, and the other contributor was busy too I think.

**Comment by Sembrani** [comment]
Yeah, I'm back at uni now, so my progress has slowed a bit. Will get back to it once I get github working properly on my PC here

**Comment by loopuleasa** [comment]
Tell me if you need any help on it.

**Comment by Sembrani** [comment]
aliases implemented, not bound to anything yet.

**Comment by loopuleasa** [comment]
I am curious where we can fit those in the main layout. They seem useful.
I haven't even had the chance to test them if they are viable or useful in comparison to default and quick.

I would experiment with replacing normal cast with smart cast.

**Comment by Sembrani** [comment]
Is this still a feature you are interested in seeing implemented or tested? I can replace normal-cast with smart-cast easily, though I haven't done a lot of testing on it and it is likely to be quite buggy

**Comment by loopuleasa** [comment]
Normal casting is good for now, but if testing proves to be better in practice, we'll probably make a different layout with smartcasting.

**Comment by loopuleasa** [comment]
invoker-orbs & toggle spells get double-tapped everytime with smartcasting?

**Comment by dodekeract** [comment]
Quite sure, yes. Since it works by using a normal + quickcast and both trigger the orb once.

**Comment by Sembrani** [comment]
An unfortunate side effect, but yes. Could be avoided by using hero keybinds, or by swapping it using ALT+SPACE+Q/W/E
